### PR TITLE
Only redraw after zooming is finished (bug 1661253)

### PR DESCRIPTION
**Pull request by calixteman**
Right now, the visible pages are redrawn for each scale change. Consequently, zooming with mouse wheel or in pinching can be pretty janky (even on a desktop machine but with a hdpi screen). So the main idea in this patch is to draw the visible pages only once zooming is finished.

### CODE REVIEW

File: web/app_options.js [hunk 1]
```diff
@@ -68,6 +68,11 @@ const defaultOptions = {
     value: 0,
     kind: OptionKind.VIEWER + OptionKind.PREFERENCE,
   },
+  defaultZoomDelay: {
+    /** @type {number */
+    value: 500,
```
Comment (Snuffleupagus) [inline web/app_options.js:73]: This seems like a quite large value, especially since you apply it to *every* zoom operation despite the commit message only talking about mouse-wheel/pinch zooming.

Looking at the changes in `web/app.js` you also (indirectly) apply it to the UI +/- buttons as well as the keyboard shortcuts, was this actually intended here!?

File: web/pdf_viewer.js [hunk 2]
```diff
@@ -2164,29 +2199,31 @@ class PDFViewer {
   /**
    * Increase the current zoom level one, or more, times.
    * @param {number} [steps] - Defaults to zooming once.
+   * @param {Object|null} [options]
    */
-  increaseScale(steps = 1) {
+  increaseScale(steps, options = null) {
```
Comment (Snuffleupagus) [inline web/pdf_viewer.js:2204]: Why are you removing the `scale = 1` default, since that leads to a number of otherwise unnecessary changes throughout the code-base?

File: web/pdf_viewer.js [hunk 3]
```diff
@@ -1080,10 +1091,18 @@ class PDFViewer {
...
+    newValue,
+    { noScroll = false, preset = false, delay = -1, update = false }
+  ) {
     this._currentScaleValue = newValue.toString();
 
     if (this.#isSameScale(newScale)) {
+      if (update) {
+        this.refresh({ scale: newScale });
+        return;
+      }
```
Comment (Snuffleupagus) [inline web/pdf_viewer.js:1105]: Based on a quick whole-words search I'm not finding anything that ever sets the `update`-option!?

**Review by Snuffleupagus — CHANGES_REQUESTED** [review]

File: web/app_options.js (hunk 1 above)
Comment (Snuffleupagus) [reply web/app_options.js:73]: I've not looked through all the code, but does that mean that if I zoom *only once* I still have to wait *half a second* for the page to actually start to re-render? Because if so, that seems like it could make the viewer feel quite sluggish; maybe just pick a lower value?

Comment (calixteman) [reply web/app_options.js:73]: In fact when you click only once on the the `+/-`, the rendering after the css zoom is not that bad and the difference between before redraw and after redraw is very tight so I'd say that the user won't see anything.

Comment (Snuffleupagus) [reply web/app_options.js:73]: > In fact when you click only once on the the `+/-`, 

That's not the only way to zoom though. Please consider that case where you first zoom to say 10% and then choose e.g. "page-width", the page will now look completely unreadable for *more than half a second*.

Comment (calixteman) [reply web/app_options.js:73]: If I understand correctly, in this case `set currentScaleValue` is used and there is no delay here so it'll behave as before this patch. Or am I missing something ?

Comment (Snuffleupagus) [reply web/app_options.js:73]: OK, that might be safe then; however I still think that a value of `500 ms` feels too high.

**Comment by calixteman** [comment]
> I'm not sure how I feel about it, and it's not really a complete patch yet, but we might be able work-around [#15812 (comment)](https://github.com/mozilla/pdf.js/pull/15812#issuecomment-1346656470) with something along the following lines[1]: [master...Snuffleupagus:pdf.js:cancel-extraDelay](https://github.com/mozilla/pdf.js/compare/master...Snuffleupagus:pdf.js:cancel-extraDelay)
> 
> [1] Where we e.g. use the `postpone`-parameter as an argument when calling `paintTask.cancel()` in the `PDFPageView` code.

After having thought about that, my feeling is that your patch is not that bad.
Right now, once zooming is finished we almost wait for 100ms to to continue or really cancel the `getOperatorList` thing because we expect that a draw could happen just once the zooming is finished.
So I'd say that we should delay the cancellation by `AppOptions.get("defaultZoomDelay") + 100`ms maybe.
In reading your comment here:
https://github.com/mozilla/pdf.js/pull/11106#issue-486803248
I think it makes sense to have two distinct values w.r.t scrolling/zooming.

**Comment by calixteman** [comment]
> It sounds like you're suggesting that the `RenderTask.cancel`-method, in the official API, should just accept a custom extra-delay. If so that feels like it could be a pretty big foot-gun, given the things that third-party users sometimes do, and we'd probably need to place an upper limit to prevent things from breaking; so how about the updated patch?
> [master...Snuffleupagus:pdf.js:cancel-extraDelay](https://github.com/mozilla/pdf.js/compare/master...Snuffleupagus:pdf.js:cancel-extraDelay)

Could we just immediately cancel when we're scrolling and just have this computed delay when zooming/rotating ?
When scrolling is there any advantages to not cancel right  now ?
I'm thinking about my own usage of the pdf viewer on my phone, and sometimes I quickly scroll the pdf to go to the page I'm looking for and in such a case I've the feeling that this delay would be a penalty.
About zooming, what do you think about postponing the cancel only when we've one page to render ? I'm wondering if it could be a penalty in general to postpone with several rendering pages.
That said, these different points could be addressed or not, in some follow-ups.
And your patch looks good for me.

**Comment by Snuffleupagus** [comment]
> Could we just immediately cancel when we're scrolling and just have this computed delay when zooming/rotating ?
> When scrolling is there any advantages to not cancel right now ?

Yes, if you quickly change scroll direction it may still make sense to slightly delay cancelling. E.g. consider a case where you scroll just past the page(s) you wanted to see and then scroll back quickly.

> About zooming, what do you think about postponing the cancel only when we've one page to render ? I'm wondering if it could be a penalty in general to postpone with several rendering pages.

I'm not sure how much of a problem that is, and anyway if there's multiple visible (or even cached) pages we really don't want to cancel their worker-thread parsing right away since that could easily regress performance (and it's somewhat difficult to imagine every possible edge-case here).
In general, I very much prefer that we don't complicate the API side of things (or even the viewer) too much regarding the cancelling. Especially given the how much things are now "moving" in this code, with the potential risk of breakage.

> That said, these different points could be addressed or not, in some follow-ups.

We could possibly try to *slightly reduce* the default cancel-delay, to e.g. `50 ms` instead, however with so many other related changes that's not really something that I think we want to do right now.

> And your patch looks good for me.

I'll submit a separate PR, however probably not today since I'm currently sick with a cold.

**Comment by calixteman** [comment]
> I'll submit a separate PR, however probably not today since I'm currently sick with a cold.

Take care of yourself.

File: web/pdf_viewer.js [hunk 4]
```diff
@@ -1099,10 +1114,24 @@ class PDFViewer {
       newScale * PixelsPerInch.PDF_TO_CSS_UNITS
     );
 
-    const updateArgs = { scale: newScale };
+    const updateArgs = {
+      scale: newScale,
+      postponeDrawingDelay: delay,
```
Comment (Snuffleupagus) [inline web/pdf_viewer.js:1119]: We should really only pass along this value if it's *actually valid*, please also see the comment below.

File: web/pdf_page_view.js [hunk 5]
```diff
@@ -564,17 +575,38 @@ class PDFPageView {
       }
     }
 
+    const postponeDrawing = postponeDrawingDelay >= 0;
```
Comment (Snuffleupagus) [inline web/pdf_page_view.js:578]: The same comments about better delay-validation, see the `pdf_viewer.js` comments, apply here as well.

File: web/pdf_page_view.js [hunk 6]
```diff
@@ -564,17 +575,38 @@ class PDFPageView {
...
+      this.textLayer?.show();
+    }
+
     if (this.canvas) {
       if (
+        postponeDrawing ||
         this.useOnlyCssZoom ||
         (this.hasRestrictedScaling && isScalingRestricted)
       ) {
+        if (postponeDrawingDelay > 0 && isFinite(postponeDrawingDelay)) {
```
Comment (Snuffleupagus) [inline web/pdf_page_view.js:592]: The same comments about better delay-validation, see `pdf_viewer.js` comments, apply here as well.

File: web/pdf_page_view.js [hunk 7]
```diff
@@ -494,7 +496,12 @@ class PDFPageView {
     }
   }
 
-  update({ scale = 0, rotation = null, optionalContentConfigPromise = null }) {
+  update({
+    scale = 0,
+    rotation = null,
+    optionalContentConfigPromise = null,
+    postponeDrawingDelay = -1,
```
Comment (Snuffleupagus) [inline web/pdf_page_view.js:503]: As mentioned previously, for overall consistency:
```suggestion
    drawingDelay = -1,
```

File: web/pdf_page_view.js [hunk 8]
```diff
@@ -654,31 +687,49 @@ class PDFPageView {
...
-    let scaleX = 1,
-      scaleY = 1;
-    if (absRotation === 90 || absRotation === 270) {
-      // Scale x and y because of the rotation.
-      scaleX = height / width;
-      scaleY = width / height;
+
+    if (target instanceof HTMLCanvasElement) {
+      if (!target.hasAttribute("zooming")) {
+        target.setAttribute("zooming", true);
```
Comment (Snuffleupagus) [inline web/pdf_page_view.js:696]: I'm not finding anything that removes this attribute, is that intentional?

File: web/pdf_page_view.js [hunk 9]
```diff
@@ -612,9 +644,10 @@ class PDFPageView {
     keepAnnotationEditorLayer = false,
     keepXfaLayer = false,
     keepTextLayer = false,
+    extraCancelDelay = 0,
```
Comment (Snuffleupagus) [inline web/pdf_page_view.js:647]: Nit: Maybe `cancelExtraDelay` instead, since it's the `extraDelay` that we're passing to the `cancel`-method?

File: web/pdf_page_view.js [hunk 10]
```diff
@@ -654,31 +687,49 @@ class PDFPageView {
     redrawAnnotationEditorLayer = false,
     redrawXfaLayer = false,
     redrawTextLayer = false,
+    postponeDrawingDelay = -1,
```
Comment (Snuffleupagus) [inline web/pdf_page_view.js:690]: It doesn't seem that this parameter is still used in the latest version of the patch.

File: web/pdf_page_view.js [hunk 11]
```diff
@@ -654,31 +687,49 @@ class PDFPageView {
...
-      this.viewport.rotation - this.paintedViewportMap.get(target).rotation;
-    const absRotation = Math.abs(relativeRotation);
-    let scaleX = 1,
-      scaleY = 1;
-    if (absRotation === 90 || absRotation === 270) {
-      // Scale x and y because of the rotation.
-      scaleX = height / width;
-      scaleY = width / height;
+
+    if (target instanceof HTMLCanvasElement) {
```
Comment (Snuffleupagus) [inline web/pdf_page_view.js:694]: Since I've not tested in detail yet, does this branch also work correctly for the `zoomLayer` (since I'm assuming it'll also hit it)?

File: web/pdf_page_view.js [hunk 12]
```diff
@@ -564,17 +575,38 @@ class PDFPageView {
       }
     }
 
+    const postponeDrawing = postponeDrawingDelay >= 0;
+
+    if (postponeDrawing) {
+      this.textLayer?.hide();
+    } else {
+      this.textLayer?.show();
+    }
```
Comment (Snuffleupagus) [inline web/pdf_page_view.js:584]: *Sorry, but I'm not following:* Why are we toggling the textLayer visibility here, since that's really not immediately clear to me?

If we actually need to do this, is there perhaps a slightly more appropriate spot to do this (e.g. in the `cancelRendering` method)?

File: web/pdf_page_view.js [hunk 13]
```diff
@@ -109,6 +109,8 @@ class PDFPageView {
     regularAnnotations: true,
   };
 
+  #previousRotation = 0;
```
Comment (Snuffleupagus) [inline web/pdf_page_view.js:112]: Nit: We've been pretty good about ordering private fields alphabetically, so can we please move this up a little bit?

File: web/pdf_viewer.js [hunk 14]
```diff
@@ -487,7 +489,16 @@ class PDFViewer {
     if (!this.pdfDocument) {
       return;
     }
-    this._setScale(val, false);
+    this._setScale(val, { noScroll: false });
+  }
+
+  setCurrentScaleValueWithOptions(val, options) {
```
Comment (Snuffleupagus) [inline web/pdf_viewer.js:495]: Do you expect any (future) direct *outside* calls of this method, since otherwise we should probably just make it private?

**Review by Snuffleupagus — CHANGES_REQUESTED** [review]

File: web/pdf_page_view.js (hunk 8 above)
Comment (calixteman) [reply web/pdf_page_view.js:696]: A new canvas will be created and will replaced the one having this attribute.
That said maybe we should just use the same canvas but it's for a follow-up I guess.

File: web/pdf_page_view.js (hunk 11 above)
Comment (calixteman) [reply web/pdf_page_view.js:694]: What do you mean exactly ?
Except in the svg case, only a canvas is passed to `cssTransform` and the `zoomLayer` is its parent.

File: web/pdf_page_view.js (hunk 12 above)
Comment (calixteman) [reply web/pdf_page_view.js:584]: The idea is to hide the text layer during the zoom session in order to not have to reflow it (which can take some time).
Here are some pictures from the profiler:
(page 1 of tracemonkey without hiding the text layer)
image
(the same but in hiding the layer)
image

File: web/pdf_page_view.js (hunk 13 above)
Comment (calixteman) [reply web/pdf_page_view.js:112]: Yep, sure.
Do you know if could add a linter for doing that automagically ?

File: web/pdf_page_view.js (hunk 12 above)
Comment (calixteman) [reply web/pdf_page_view.js:584]: After thinking about that, I'm not sure it's a good idea to move that stuff in `cancelRendering` because we don't want to cancel something here and it doesn't make to hide it in the non-postpone case.

File: web/pdf_page_view.js (hunk 8 above)
Comment (calixteman) [reply web/pdf_page_view.js:701]: The idea behind this change is just to move the rescale stuff in the css to not have to change something in using js.

File: web/pdf_page_view.js (hunk 11 above)
Comment (Snuffleupagus) [reply web/pdf_page_view.js:699]: I was just wondering if the zoomLayer is "special" in some way, since I wasn't entirely sure how it worked (it was a long time since I looked at its code).

File: web/pdf_page_view.js [hunk 15]
```diff
@@ -569,17 +580,39 @@ class PDFPageView {
...
+        }
+
         this.cssTransform({
           target: this.canvas,
           redrawAnnotationLayer: true,
           redrawAnnotationEditorLayer: true,
           redrawXfaLayer: true,
-          redrawTextLayer: true,
+          redrawTextLayer: !postponeDrawing,
+          postponeDrawingDelay,
```
Comment (Snuffleupagus) [inline web/pdf_page_view.js:615]: This parameter is no longer used in that method, and can thus be removed here.

File: web/pdf_viewer.js [hunk 16]
```diff
@@ -1099,10 +1114,27 @@ class PDFViewer {
       newScale * PixelsPerInch.PDF_TO_CSS_UNITS
     );
 
-    const updateArgs = { scale: newScale };
+    const mustPostponeDrawing = drawingDelay >= 0 && drawingDelay < 1000;
+    const updateArgs = {
+      scale: newScale,
+    };
+    if (mustPostponeDrawing) {
+      updateArgs.postponeDrawingDelay = drawingDelay;
```
Comment (Snuffleupagus) [inline web/pdf_viewer.js:1122]: Again, can't we just call the parameter `drawingDelay` consistently everywhere?
```suggestion
      updateArgs.drawingDelay = drawingDelay;
```

File: web/pdf_viewer.js [hunk 17]
```diff
@@ -2098,11 +2134,10 @@ class PDFViewer {
     this.#annotationEditorUIManager.updateParams(type, value);
   }
 
-  refresh() {
+  refresh(updateArgs = Object.create(null)) {
     if (!this.pdfDocument) {
       return;
     }
-    const updateArgs = {};
```
Comment (Snuffleupagus) [inline web/pdf_viewer.js:2105]: These changes are no longer needed, as far as I can tell.

File: web/pdf_viewer.js [hunk 18]
```diff
@@ -1099,10 +1114,27 @@ class PDFViewer {
...
+    if (mustPostponeDrawing) {
+      updateArgs.postponeDrawingDelay = drawingDelay;
+    }
     for (const pageView of this._pages) {
       pageView.update(updateArgs);
     }
+
+    if (this.#scaleTimeoutId !== null) {
+      clearTimeout(this.#scaleTimeoutId);
+    }
```
Comment (Snuffleupagus) [inline web/pdf_viewer.js:1130]: Do we perhaps want to try and reset this timeout in more places as well, I'm thinking in particular about the `PDFViewer.update` method?

Consider the following, somewhat contrived, series of events:
 1. The user modifies the new preference to its maximum value.
 2. The user zoom in/out quickly a few times, using either a mouse or keyboard shortcuts.
 3. The user then *immediately* rotates the document.

In this case we immediately re-render the rotated document, however the `#scaleTimeoutId` may still be running and it could thus force us to almost immediately invalidate/re-render again.

Note here that in this case, since the scale is a numerical value and because of the `if (this.#isSameScale(newScale)) { ... }`-branch above, we'd not actually reset the timeout in this case.

**Review by Snuffleupagus — COMMENTED** [review]
This is looking much better now, but I've got a few more small suggestions/questions here.

**Review by Snuffleupagus — COMMENTED** [review]
*Submitted twice by mistake.*

File: web/pdf_page_view.js [hunk 19]
```diff
@@ -569,17 +580,35 @@ class PDFPageView {
...
+
     if (this.canvas) {
       if (
+        postponeDrawing ||
         this.useOnlyCssZoom ||
         (this.hasRestrictedScaling && isScalingRestricted)
       ) {
+        if (
+          postponeDrawing &&
+          this.renderingState === RenderingStates.RUNNING
```
Comment (Snuffleupagus) [inline web/pdf_page_view.js:593]: Is this check actually sufficient, since there's also the `PAUSED` state (where it also makes sense to cancel rendering)?
Do we perhaps just want to do this whenever `this.renderingState !== RenderingStates.FINISHED` instead? Obviously that'd include the `INITIAL` state, but given that no rendering happens in that case rendering-cancellation becomes a no-op then.

File: web/pdf_viewer.js [hunk 20]
```diff
@@ -1099,10 +1112,22 @@ class PDFViewer {
...
+    const mustPostponeDrawing = drawingDelay >= 0 && drawingDelay < 1000;
+    const updateArgs = {
+      scale: newScale,
+    };
+    if (mustPostponeDrawing) {
+      updateArgs.drawingDelay = drawingDelay;
+    }
+    this.refresh(false, updateArgs);
+
+    if (mustPostponeDrawing) {
```
Comment (Snuffleupagus) [inline web/pdf_viewer.js:1124]: With the delayed rendering we're no longer toggling the `loadingIcon`-div on the pages until the actual re-rendering starts. It feels a little surprising to re-size the page, using the CSS-only zooming, and then *after* the `drawingDelay` having the indicator suddenly show up.

Don't we still want to keep showing the `loadingIcon`-div, like in the current code, immediately when rendering starts?
If so, we could probably fix that here by adding something like 
```suggestion
    if (mustPostponeDrawing) {
      const visiblePages = this._getVisiblePages();
      this.#toggleLoadingIconSpinner(visiblePages.ids);
      
```

**Review by Snuffleupagus — COMMENTED** [review]
This is *so* close to done now, however I found a few small issue in the latest update (see the inline comments).

File: web/pdf_page_view.js (hunk 19 above)
Comment (calixteman) [reply web/pdf_page_view.js:593]: I managed to have an exception (not easy to reproduce: I reload and zoom up/down before loading has finished) because we're cancelling a paused task and then at some point `resume` was called but it has been nulled in `cancelRendering`:
```
Uncaught (in promise) TypeError: view.resume is not a function
    renderView http://localhost:8888/web/pdf_rendering_queue.js:189
    forceRendering http://localhost:8888/web/pdf_viewer.js:1672
    promise callback*forceRendering http://localhost:8888/web/pdf_viewer.js:1671
    renderHighestPriority http://localhost:8888/web/pdf_rendering_queue.js:80
    update http://localhost:8888/web/pdf_viewer.js:1464
    _scrollUpdate http://localhost:8888/web/pdf_viewer.js:1039
    viewAreaElementScrolled http://localhost:8888/web/ui_utils.js:178
    debounceScroll http://localhost:8888/web/ui_utils.js:163
    watchScroll http://localhost:8888/web/ui_utils.js:191
    PDFViewer http://localhost:8888/web/pdf_viewer.js:310
    _initializeViewerComponents http://localhost:8888/web/app.js:482
    initialize http://localhost:8888/web/app.js:239
    run http://localhost:8888/web/app.js:643
    webViewerLoad http://localhost:8888/web/viewer.js:222
    promise callback*webViewerLoad http://localhost:8888/web/viewer.js:221
    <anonymous> http://localhost:8888/web/viewer.js:258
```
And if I add `this.renderingState = RenderingStates.INITIAL;` after the `cancelRendering` like we do in `reset`, sometimes the space between two pages disappears during zooming. I think it'd make sense to reset the state.
I'll have a look on that stuff tomorrow, but if you've any ideas I take.

Comment (Snuffleupagus) [reply web/pdf_page_view.js:593]: With the caveat that I've gotten sick again and possible isn't thinking straight; isn't there potentially a "bigger picture" issue here?

If the page isn't `FINISHED` rendering when we get to here, won't the next call to `PDFViewer.update` (which happens e.g. via the "scalechanging" event listener) actually force it to start re-rendering *immediately* rather than being delayed?
Hence it seems to me that we'd need a way to both cancel any pending rendering here *and* also ensure that it won't attempt to start re-rendering before delay actually runs out.

Comment (Snuffleupagus) [reply web/pdf_page_view.js:593]: Actually, do we perhaps just want to set `this.renderingState = RenderingStates.FINISHED;` regardless here to avoid this?

File: web/pdf_viewer.js (hunk 20 above)
Comment (calixteman) [reply web/pdf_viewer.js:1124]: In general, I'd be rather in favor of just displaying the loading icon only when the page redrawing is taking more than 1 second (or something like that) just to inform the user that the viewer is not stuck. 
I don't really see how this spinner could be useful for the user especially when they're zooming.

File: web/pdf_page_view.js (hunk 19 above)
Comment (calixteman) [reply web/pdf_page_view.js:593]: I was adding a new state `POSTPONED` but I think you're right and using `FINISHED` is good enough.

File: extensions/chromium/preferences_schema.json [hunk 21]
```diff
@@ -28,6 +28,12 @@
       ],
       "default": 0
     },
+    "defaultZoomDelay": {
+      "title": "Default zoom delay",
+      "description": "Delay (in ms) to wait before redrawing the canvas.",
+      "type": "integer",
+      "default": 500
```
[the remark this answers is not in the document]
Comment (Snuffleupagus) [reply extensions/chromium/preferences_schema.json:35]: This still needs to be updated :-)

File: web/pdf_viewer.js [hunk 22]
```diff
@@ -2098,15 +2123,19 @@ class PDFViewer {
     this.#annotationEditorUIManager.updateParams(type, value);
   }
 
-  refresh() {
+  refresh(update, updateArgs = Object.create(null)) {
```
Comment (Snuffleupagus) [inline web/pdf_viewer.js:2126]: This is actually a breaking change, so please adjust this as below and update the call-sites accordingly.
```suggestion
  refresh(noUpdate = false, updateArgs = Object.create(null)) {
```

File: web/pdf_viewer.js [hunk 23]
```diff
@@ -2098,15 +2123,19 @@ class PDFViewer {
...
     if (!this.pdfDocument) {
       return;
     }
-    const updateArgs = {};
     for (const pageView of this._pages) {
       pageView.update(updateArgs);
     }
-    this.update();
+    if (this.#scaleTimeoutId !== null) {
+      clearTimeout(this.#scaleTimeoutId);
```
Comment (Snuffleupagus) [inline web/pdf_viewer.js:2134]: Let's null it as well, to prevent this code-path from running more than once.
```suggestion
      clearTimeout(this.#scaleTimeoutId);
      this.#scaleTimeoutId = null;
```

File: web/pdf_viewer.js (hunk 22 above)
Comment (Snuffleupagus) [reply web/pdf_viewer.js]: This still requires follow-up adjustments to all call-sites, otherwise I don't thing that things will work :-)

File: web/pdf_page_view.js [hunk 24]
```diff
@@ -245,6 +247,7 @@ class PDFPageView {
       scale: this.scale * PixelsPerInch.PDF_TO_CSS_UNITS,
       rotation: totalRotation,
     });
+    this.#setDimensions();
```
Comment (Snuffleupagus) [inline web/pdf_page_view.js:250]: Do you perhaps also want to set `this.#previousRotation = totalRotation;` here, to ensure that the value is always correct?
It seems to me that that could *potentially* avoid having to re-run `#setDimensions` is some more cases in the `update`-method.

*Ninja-edit:* Or possible just add something like `this.#previousRotation = viewport.rotation;` at the top of the `#setDimensions`-method, to avoid having to do that *manually* in multiple spots?

File: web/pdf_viewer.js [hunk 25]
```diff
@@ -2005,29 +2028,31 @@ class PDFViewer {
...
-  increaseScale(steps = 1) {
+  increaseScale(steps = 1, options = null) {
     let newScale = this._currentScale;
     do {
       newScale = (newScale * DEFAULT_SCALE_DELTA).toFixed(2);
       newScale = Math.ceil(newScale * 10) / 10;
       newScale = Math.min(MAX_SCALE, newScale);
     } while (--steps > 0 && newScale < MAX_SCALE);
-    this.currentScaleValue = newScale;
+    this.#setCurrentScaleValueWithOptions(newScale, options);
```
Comment (Snuffleupagus) [inline web/pdf_viewer.js:2040]: Do we actually need the new helper method, or could we simply inline the necessary code as below?
```suggestion
    options ||= Object.create(null);
    options.noScroll = false;
    this._setScale(newScale, options);
```

File: web/pdf_viewer.js [hunk 26]
```diff
@@ -2005,29 +2028,31 @@ class PDFViewer {
...
-  decreaseScale(steps = 1) {
+  decreaseScale(steps = 1, options = null) {
     let newScale = this._currentScale;
     do {
       newScale = (newScale / DEFAULT_SCALE_DELTA).toFixed(2);
       newScale = Math.floor(newScale * 10) / 10;
       newScale = Math.max(MIN_SCALE, newScale);
     } while (--steps > 0 && newScale > MIN_SCALE);
-    this.currentScaleValue = newScale;
+    this.#setCurrentScaleValueWithOptions(newScale, options);
```
Comment (Snuffleupagus) [inline web/pdf_viewer.js:2055]: Do we actually need the new helper method, or could we simply inline the necessary code as below?
```suggestion
    options ||= Object.create(null);
    options.noScroll = false;
    this._setScale(newScale, options);
```

File: web/text_layer_builder.js [hunk 27]
```diff
@@ -101,10 +101,10 @@ class TextLayerBuilder {
           mustRescale,
           mustRotate,
         });
-        this.show();
         this.#scale = scale;
         this.#rotation = rotation;
       }
+      this.show();
```
Comment (Snuffleupagus) [inline web/text_layer_builder.js:107]: This is a really good find, since it actually fixes an existing bug that I've seen intermittently!
(Where the textLayer sometimes would never become visible again after zooming, if you changed zoom-level multiple times for a slowly rendering document).

File: web/app_options.js [hunk 28]
```diff
@@ -68,6 +68,11 @@ const defaultOptions = {
     value: 0,
     kind: OptionKind.VIEWER + OptionKind.PREFERENCE,
   },
+  defaultZoomDelay: {
+    /** @type {number} */
+    value: 400,
```
Comment (Snuffleupagus) [inline web/app_options.js:73]: Given that the next release should happen fairly soon, we might not want to enable the zooming-part of the patch just before that (especially with many people now being on holiday) to have more time to find and fix any potential regressions.

However if you want to land this PR *right away*, since it contains a bunch of re-factoring of existing method signatures, I'd simply suggest that we *temporarily* disable the zooming-part until after the release like so:
```suggestion
    value: typeof PDFJSDev === "undefined" || !PDFJSDev.test("GENERIC") ? 400 : -1,
```

**Review by Snuffleupagus — APPROVED** [review]
Sorry about the delay in getting through this, but between being sick (multiple times) and the size/scope of the changes this was quite challenging to review.

Hopefully I've not missed anything obvious here, since I've spent quite a lot of time thinking about the "surrounding" code. However with the new preference we have a simple way to disable this functionality, should that ever become necessary.

r=me, with a couple of final comments; thank you!

File: web/text_layer_builder.js (hunk 27 above)
Comment (calixteman) [reply web/text_layer_builder.js:107]: Yep, my way to test zoom+text_layer is to ctrl+a and zoom to see if everything is ok.
